var/spack/repos/builtin/packages/chapel/package.py

```diff
@@ -561,10 +588,74 @@ def unset_chpl_env_vars(self, env):
         for var in self.chpl_env_vars:
             env.unset(var)
 
+    def get_path(self, s: str) -> typing.List[str]:
+        get_path_pat = re.compile(r"(?:RUNPATH|RPATH)=(.*)")
+        m = re.search(get_path_pat, s)
+        if m:
+            paths = m.group(1).split(":")
+            paths = [p.strip() for p in paths if p.strip() != ""]
+            return paths
+        return []
+
+    # Some chapel build scripts depend on CHPL_HOME, this can get encoded into
+    # the frontend library rpaths. this is problematic for the python
+    # bindings which get installed in virtualenv provided by spack. This script
+    # will fix the rpath of the chapel frontend library after installation
+    @run_after("install")
+    def fix_rpath(self):
+        files = []
+        # find all .so in the old lib path
+        for f in glob.glob(
+            join_path(self.prefix.lib, "chapel", str(self._output_version_short), "**", "*.so"),
+            recursive=True,
+        ):
+            files.append(f)
+
+        for f in glob.glob(
+            join_path(self.spec["python-venv"].prefix.lib, "**", "chapel", "*.so"), recursive=True
+        ):
+            files.append(f)
+
+        for f in files:
+            cp = subprocess.run(
+                ["chrpath", "-l", f],
+                encoding="utf-8",
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+            )
+            path = cp.stdout.strip()
+            path = self.get_path(path)
+            new_path = []
+            # clean path
+            for p in path:
+                if p == self.prefix.lib64:
+                    continue
+                if p.startswith(str(self.build_directory)):
+                    new_path.append(
+                        str(
+                            join_path(
+                                self.prefix.lib, "chapel", self._output_version_short, "compiler"
```

Comment:
This whole procedure seems pretty hack-ish, and I don't really understand why it's necessary.

RPATH entries listing inaccessible or non-existent directories (like the build dir after install) _should_ be silently ignored by the Linux loader. As such, removing bogus entries should be unnecessary (unless they happen to exist and contain a conflicting library of the same name). Did you find that was not the case?

It seems to me like the only RPATH change this procedure is _actually_ making that should affect behavior is the removal of `$install/lib64` (which I suspect is incidental?) and insertion of `$install/lib/chapel/$ver/compiler`. If that is the actual point of this routine, then please update the block comment above to explain that's what this code is actually trying to accomplish.

I also question whether this is the best way to accomplish the needed change. Can we not instead inject the missing necessary RPATH entries into linker options before the build? (ie by arranging for the Chapel equivalent of `LDFLAGS+=-Wl,-rpath=$install/lib/chapel/$ver/compiler` ) That pre-link solution also seems safer, since IIUC the current procedure could fail at `chrpath` time if `self.prefix` is significantly longer than `self.build_directory`...

Reply:
This came from adding the python-bindings. The point is to be able to import the chapel bindings in python. Without changing the `RPATH` we have to add an `LD_LIBRARY_PATH` entry or else we'll get errors in Python when trying to import chapel. 

>Can we not instead inject the missing necessary RPATH entries into linker options before the build?

It may be possible to do this by patching the `setup.py` file in this location:
```
LDFLAGS += [
    "-L{}".format(chpl_lib_path),
    "-lChplFrontendShared",
    "-Wl,-rpath",
    chpl_lib_path,
]
```
such that we rewrite the `chpl_lib_path` before it's written to the lib's rpath.

Reply:
ok, the idea of injecting the proper rpath at build time failed because the path doesn't exist until after the installation

Reply:
> rewrite the chpl_lib_path before it's written to the lib's rpath ...
> idea of injecting the proper rpath at build time failed because the path doesn't exist until after the installation

Did you attempt _replacing_ the build rpath with the install rpath, or _appending_ the install rpath? I was suggesting the latter, so that we provide both directories in the rpath, which should hopefully ensure it works in pre- or post- install contexts.